stdlib/public/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.cpp

```diff
@@ -0,0 +1,2444 @@
+    std::optional<std::unique_ptr<llvm::MemoryBuffer>> &&memoryBuffer,
+    const std::string &path) {
+  if (isa<llvm::object::MachOObjectFile>(binary.get())) {
+    auto objectFile = cast<llvm::object::MachOObjectFile>(std::move(binary));
+    auto filetype = objectFile->getHeader().filetype;
+    // TODO: executables, bundles?
+    if (filetype != llvm::MachO::MH_DYLIB)
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "Unsupported mach-o filetype %" PRIu32,
+                                     filetype);
+    LOG(LogLevel::Info, "Loaded %p from %s", objectFile.get(), path.c_str());
+    MachOFile *file =
+        new MachOFile{std::move(memoryBuffer), std::move(objectFile), path};
+    addImage(file);
+    return llvm::Error::success();
+  }
+
+  if (auto universal =
+          dyn_cast<llvm::object::MachOUniversalBinary>(binary.get())) {
+    auto objectForArch = universal->getMachOObjectForArch(arch);
+    if (!objectForArch)
+      return objectForArch.takeError();
+    auto objectFile = objectForArch.get().get();
+    LOG(LogLevel::Info, "Loaded %p from %s", objectFile, path.c_str());
+    MachOFile *file = new MachOFile{std::move(memoryBuffer),
+                                    std::move(objectForArch.get()), path};
+    addImage(file);
+    return llvm::Error::success();
+  }
+
+  return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                 "Unknown binary type.");
+}
+
+template <typename Runtime>
+llvm::Error ExternalGenericMetadataBuilderContext<Runtime>::addImageAtPath(
+    const std::string &path) {
+  LOG(LogLevel::Info, "Adding %s", path.c_str());
+
+  auto binaryOwner = llvm::object::createBinary(path);
+  if (!binaryOwner) {
+    return binaryOwner.takeError();
+  }
+  auto [binary, buffer] = binaryOwner->takeBinary();
+  return addImage(std::move(binary), std::optional(std::move(buffer)), path);
+}
+
+template <typename Runtime>
+llvm::Error ExternalGenericMetadataBuilderContext<Runtime>::addImageInMemory(
+    const void *start, uint64_t length, const std::string &path) {
+  LOG(LogLevel::Info, "Adding %s in memory", path.c_str());
+  llvm::StringRef stringRef{reinterpret_cast<const char *>(start), length};
+  llvm::MemoryBufferRef bufferRef{stringRef, {}};
+  auto binary = llvm::object::createBinary(bufferRef);
+  if (!binary)
+    return binary.takeError();
+  return addImage(std::move(binary.get()), {}, path);
+}
+
+/// Add images in the given path. Returns an error if there was a filesystem
+/// error when iterating over the directory. Errors reading individual images
+/// are NOT returned, as it's assumed that there may be non-dylib content in the
+/// directory as well.
+template <typename Runtime>
+BuilderErrorOr<std::monostate>
+ExternalGenericMetadataBuilderContext<Runtime>::addImagesInPath(
+    std::string path) {
+  auto forEachFile = [&](std::string filepath) {
+    auto error = addImageAtPath(filepath);
```

Comment:
I'm not sure what the right approach is here. It doesn't matter _too_ much, as `swift_externalMetadataBuilder_addDylib` is intended to be the primary interface, and this is a convenience. For the convenience interface, I'd rather it try load everything in sight rather than have it potentially skip something we want to keep if somebody puts a weird extension on their dylib.